### PR TITLE
Use the public `on` method when listening

### PR DESCRIPTION
**Pull request by jridgewell**
Fixes https://github.com/jashkenas/backbone/issues/3611.

This uses a private `listening` var to share state between a Backbone
"listener" and "listenee", instead of using a private `internalOn()` to
share state. This allows `#listenTo` to use the public `#on` method and
keeps interop between Backbone and any other event library.

http://jsperf.com/internal-listening-public-on

### CODE REVIEW

File: backbone.js [hunk 1]
```diff
@@ -149,10 +146,20 @@
     }
 
     // Bind callbacks on obj, and keep track of them on listening.
-    internalOn(obj, name, callback, this, listening);
+    var error = tryCatchOn(obj, name, callback, this);
+    listening = void 0;
```
Comment (megawac) [inline backbone.js]: Why is this necessary - won't it be gced anyway

Comment (jridgewell) [inline backbone.js]: It's [now](https://github.com/jashkenas/backbone/pull/3615/files#diff-0d56d0d310de7ff18b3cef9c2f8f75dcR91) a private global.

**Comment by jridgewell** [comment]
Hold off on this for today, it has an issue because the other library won't update `listening.count`, so we don't know when we can cleanup references to `listenee` from the `listener`.

**Comment by jashkenas** [comment]
"The other library" ?

Also, `tryCatchOn` isn't a good thing.

**Comment by jridgewell** [comment]
I've just revised the implementation to use a hybrid approach between #3455 and #3594.

If the listenee is using Backbone's `#on`, the `listening` will keep track of callbacks using a fast counter (#3594's approach). If not, then callbacks will be stored in an `_events` object on the `listening` object (#3455). This allows full interoperability while also properly cleaning up memory bindings when all callbacks have been removed.

[JsPerf](http://jsperf.com/internal-listening-public-on/4) tends to like the implementation, except Chrome doesn't like `#listenTo` (running the same benchmark in node v0.12 shows a less than 0.2% slowdown).

- - -

> Also, tryCatchOn isn't a good thing.

Can you explain?

**Comment by DomBlack** [comment]
Is there any chance this PR will be merged?

**Comment by alexbaumgertner** [comment]
@DomBlack 👍  
I'm interested in this code too. 
This PR can fix at least one painful bug in my project. 
Do the author need any help?

**Comment by alexbaumgertner** [comment]
@jridgewell I tried to rebase and fix merge errors in https://github.com/jashkenas/backbone/pull/4043 but a bunch of tests fails. Could you help me?

**Comment by jbboehr** [comment]
Any updates on this? I've merged master into it in #4131, if that helps at all.
